### PR TITLE
Add support for [GenerateImmutable] on nested classes. Fixes #52.

### LINKED ISSUE #52: [GenerateImmutable] on nested classes generates code with errors.
Applying `[GenerateImmutable]` to a nested class generates code that won't compile. For example:

```c#
    partial class Nested
    {
        [GenerateImmutable( GenerateBuilder = true, DefineWithMethodsPerProperty = true )]
        public partial class NestedClass
        {
            readonly string name;
        }
    }
```

### PR DESCRIPTION
**Pull request by tg73**
Fixes #52

### CODE REVIEW

**Comment by AArnott** [comment]
Thanks for contributing. I love that you included tests for the change as well. I'll review soon and merge or send some feedback.

**Comment by tg73** [comment]
Hi Andrew - thanks for the comment. This is my first brush with Roslyn, so fingers crossed.

I've realised that there's a flaw in the namespace processing, which applies with our without my changes, to any class, nested or otherwise. Consider the following:
```
namespace X { namespace Y { [GenerateImmutable] partial class Z {... } } }
```
IOG will 'see' only the `Y` namespace, emitting a partial class like:
```
namespace Y { [GenerateImmutable] partial class Z {... } }
```
I was having a quick look at how to address this: from what I understand of Roslyn (remember, I'm a newbie), at present the namespace is determined by traversing the syntax tree. From `DocumentTransform.cs`:
```
var namespaceNode = memberNode.Parent as NamespaceDeclarationSyntax;
```
which I changed to:
```
var namespaceNode = memberNode.Ancestors().OfType<NamespaceDeclarationSyntax>().FirstOrDefault();
```
Would it instead be better to obtain the namespace of `memberNode` by using the semantic model, for example something based on:
```
inputSemanticModel.GetTypeInfo( memberNode ).Type.ContainingNamespace
```
Your thoughts? It would be useful to understand your reasoning as applies throughout IOG for when you prefer syntax tree over semantic model and vice versa.

Tom

**Comment by AArnott** [comment]
Thanks, Tom. We can fix the nested namespace issue in a separate pull request and go with your present fix to solve the nested class problem. 
I think I prefer the syntax tree API more than semantic model where either one works equally well. But when the semantic model makes things more straightforward and avoids deeper understanding of C#, then by all means, lets use the semantic model.

File: src/ImmutableObjectGraph.CodeGeneration.Roslyn/CodeGen.cs [hunk 1]
```diff
@@ -166,6 +166,18 @@ private void MergeFeature(FeatureGenerator featureGenerator)
 
             partialClass = this.mergedFeatures.Aggregate(partialClass, (acc, feature) => feature.ProcessApplyToClassDeclaration(acc));
             var outerMembers = SyntaxFactory.List<MemberDeclarationSyntax>();
+            if ( applyTo.Parent.Kind() == SyntaxKind.ClassDeclaration )
+            {
+                var parent = applyTo.Parent as ClassDeclarationSyntax;
+
+                while ( parent != null )
```
Comment (AArnott) [inline src/ImmutableObjectGraph.CodeGeneration.Roslyn/CodeGen.cs]: Your whitespace rules don't match VS defaults. Can you adjust this to match the surrounding code? For instance, no space just inside your parentheses.

**Comment by AArnott** [comment]
The changes look good, once the style issues are fixed up.

**Comment by tg73** [comment]
Hi Andrew - this issue of whitespace settings is one area where VS is not very helpful. I'm not aware of an easy way to have different settings for different solutions or repos. Changing too and fro between defaults and our team settings where I work is pretty awkward - if I simply turn off team settings, VS does not return to defaults - so I would have to explicitly reset everything, perhaps losing my own personal customisations. Then there's the issue of having multiple solutions open in multiple VS instances (which I often do), with different settings in different instances. Do you have any suggestion for dealing with this? In the mean time, I've been trying to do it by hand, but clearly missed a few cases, which I'll correct now.

Clearly I'm not alone in this frustration:

https://visualstudio.uservoice.com/forums/121579-visual-studio/suggestions/2089769-store-per-project-source-formatting-settings-with

**Comment by IanYates** [comment]
@tg73 - I've come across these pages in the past.  I've just searched and found them again

https://vsswitch.codeplex.com/
https://visualstudiogallery.msdn.microsoft.com/410e9b9f-65f3-4495-b68e-15567e543c58

(There's also "EditorConfig" but from what I can read Rebracer seems to be more VS friendly)
Summary pages I found
http://stackoverflow.com/questions/1212092/visual-studio-per-solution-indentation-settings
http://visualstudioextensions.vlasovstudio.com/2013/12/30/visual-studio-code-formatting-options-per-solution-directory-and-file/

I haven't tried either myself...

**Comment by AArnott** [comment]
I've struggled with this as well, and mostly settled for adjusting team settings to the VS defaults to mitigate the pain. But while the variance between these two lasts, you can Save Settings and Restore Settings to switch between the two modes. 
But I'm afraid when contributing to most open source projects, it's something you'll probably have to get used to as most open source project owners I've conversed with expect you to follow their rules, whatever they may be (usually VS defaults).

**Comment by AArnott** [comment]
It looks good with your fix commit though. I'll merge. Thanks again for the contribution.

**Comment by tg73** [comment]
Re contribution - you're welcome. And thanks to you for the project as a whole. Also may I give you a thumbs up for your prompt, clear and constructive handling of my pull requests.

Re VS settings - thanks for the various comments/ideas on this. I'll give the save/restore settings idea a go, as it seems the least intrusive and doesn't require custom extensions.

**Pull request merged**